src/main/java/carpet/script/api/Scoreboards.java

```diff
@@ -147,5 +154,295 @@ public static void apply(Expression expression)
             scoreboard.setObjectiveSlot(slot, objective);
             return (_c, _t) -> new NumericValue(slot);
         });
+
+        expression.addLazyFunction("team_list", -1, (c, t, lv) ->
+        {
+            if(lv.size() > 1) throw new InternalExpressionException("'team_list' requires zero or one parameters");
+            CarpetContext cc = (CarpetContext)c;
+            ServerScoreboard scoreboard = cc.s.getMinecraftServer().getScoreboard();
+            if(lv.size() == 0) {
+                Value ret = ListValue.wrap(scoreboard.getTeamNames().stream().map(StringValue::new).collect(Collectors.toList()));
+                return (_c, _t) -> ret;
+            } else if(lv.size() == 1) {
+                Value teamVal = lv.get(0).evalValue(c);
+                if(teamVal instanceof StringValue) {
+                    Team team = scoreboard.getTeam(lv.get(0).evalValue(c).getString());
+                    if(team == null) return LazyValue.NULL;
+                    Value ret = ListValue.wrap(team.getPlayerList().stream().map(StringValue::new).collect(Collectors.toList()));
+                    return (_c, _t) -> ret;
+                } else {
+                    throw new InternalExpressionException("'team_list' requires a string as the first argument");
+                }
+            } else {
+                return LazyValue.NULL;
+            }
+        });
+
+
+        expression.addLazyFunction("team_add", -1, (c, t, lv) ->
+        {
+            if(!(lv.size() < 3 && lv.size() > 0)) throw new InternalExpressionException("'team_add' requires one or two parameters");
+
+            CarpetContext cc = (CarpetContext)c;
+            ServerScoreboard scoreboard = cc.s.getMinecraftServer().getScoreboard();
+            Value teamVal = lv.get(0).evalValue(c);
+            if(!(teamVal instanceof StringValue)) throw new InternalExpressionException("'team_add' requires a string as the first argument");
+
+            String teamName = teamVal.getString();
+
+            if(lv.size() == 1) {
+                if(scoreboard.getTeam(teamName) == null) {
+                    scoreboard.addTeam(teamName);
+                    Value ret = new StringValue(teamName);
+                    return (_c, _t) -> ret;
+                } else {
+                    return LazyValue.NULL;
+                }
+            } else if(lv.size() == 2) {
+                Value playerVal = lv.get(1).evalValue(c);
+                String player;
+                if(playerVal instanceof StringValue) {
```

Comment:
EntityValue::getPlayerByValue should handle all the cases for you - unless you want to cover usecases where its a non-logged player. In that case I suggest adding a new function that looks similar EntityValue::getPlayerNameByValue that does the same thing, but honors nonexisting players and return String, not the ServerPlayerEntity instead. - that function would be handy in several other places.